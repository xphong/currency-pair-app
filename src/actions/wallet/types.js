export const REQUEST_TRADE = `wallet/REQUEST_TRADE`;
export const COMPLETE_TRADE = `wallet/COMPLETE_TRADE`;
export const TRADE_ERROR = `wallet/TRADE_ERROR`;
