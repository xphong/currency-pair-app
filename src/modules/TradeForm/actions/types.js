export const REQUEST_QUOTE = `quote/REQUEST_QUOTE`;
export const RETRIEVE_QUOTE = `quote/RETRIEVE_QUOTE`;
export const QUOTE_ERROR = `quote/QUOTE_ERROR`;
