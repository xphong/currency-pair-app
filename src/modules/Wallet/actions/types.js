export const REQUEST_TRADE = `wallet/REQUEST_TRADE`;
export const RECEIVE_TRADE = `wallet/RECEIVE_TRADE`;
export const TRADE_ERROR = `wallet/TRADE_ERROR`;
