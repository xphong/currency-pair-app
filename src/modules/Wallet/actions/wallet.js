import * as types from './types';

import delay from 'utils/delay/delay';

export const requestTrade = () => {
  return {
    type: types.REQUEST_TRADE
  };
}

export const receiveTrade = balance => {
  return {
    type: types.RECEIVE_TRADE,
    payload: {
      balance
    }
  };
}

export const receiveError = error => {
  return {
    type: types.TRADE_ERROR,
    payload: {
      error
    }
  };
}

export const trade = (amount, quoteAmount) => {
  return async (dispatch, getState) => {
    const { wallet } = getState();
    const usdBalance = wallet.balance.USD;
    const btcBalance = wallet.balance.BTC;

    if (!amount || !quoteAmount) {
      dispatch(receiveError('Please enter a trade amount.'));
      return;
    }

    if (amount > usdBalance) {
      dispatch(receiveError('You have insufficient funds in your account.'));
      return;
    }

    dispatch(requestTrade());

    try {
      // Fake a delay. In a real-world application, trade would make a POST call
      await delay(1000);

      const balance = {
        USD: usdBalance - amount,
        BTC: btcBalance + quoteAmount
      };

      dispatch(receiveTrade(balance));
    } catch (e) {
      dispatch(receiveError(e.message));
    }
  };
}
