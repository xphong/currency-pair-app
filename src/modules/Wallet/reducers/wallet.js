import * as types from 'modules/Wallet/actions/types';

const INITIAL_STATE = {
  isLoading: false,
  error: false,
  success: false,
  balance: {
    USD: 156.12,
    BTC: 0
  }
};

const wallet = (state = INITIAL_STATE, action = null) => {
  switch (action.type) {
    case types.REQUEST_TRADE:
      return Object.assign({}, state, { isLoading: true, error: false, success: false });
    case types.TRADE_ERROR:
      return Object.assign({}, state, {
        isLoading: false,
        error: action.payload.error,
        success: false
      });
    case types.RECEIVE_TRADE:
      return Object.assign({}, state, {
        isLoading: false,
        balance: action.payload.balance,
        error: false,
        success: true
      });
    default:
      return state;
  }
}

export default wallet;
