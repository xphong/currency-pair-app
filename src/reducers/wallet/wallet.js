import * as types from '../../actions/wallet/types';

const INITIAL_STATE = {
  isLoading: false,
  error: false,
  balance: {
    USD: 156.12,
    BTC: 0
  }
};

const wallet = (state = INITIAL_STATE, action = null) => {
  switch (action.type) {
    case types.REQUEST_TRADE:
      return Object.assign({}, state, { isLoading: true, error: false });
    case types.TRADE_ERROR:
      return Object.assign({}, state, {
        isLoading: false,
        balance: action.payload.balance,
        error: true
      });
    case types.COMPLETE_TRADE:
      return Object.assign({}, state, {
        isLoading: false,
        balance: action.payload.balance,
        error: false
      });
    default:
      return state;
  }
}

export default wallet;
