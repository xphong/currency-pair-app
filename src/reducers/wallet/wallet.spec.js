import * as types from '../../actions/wallet/types';
import walletReducer from './wallet';

describe('Wallet Reducer', () => {
  it('should return initial state', () => {
    const reducer = walletReducer(undefined, {});

    expect(reducer).toMatchSnapshot();
  });

  it('should handle REQUEST_TRADE', () => {
    const action = {
      type: types.REQUEST_TRADE
    };

    const reducer = walletReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });

  it('should handle TRADE_ERROR', () => {
    const action = {
      type: types.TRADE_ERROR,
      payload: {
        error: 'Insuccifient funds'
      }
    };

    const reducer = walletReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });

  it('should handle RECEIVE_TRADE', () => {
    const action = {
      type: types.RECEIVE_TRADE,
      payload: {
        balance: {
          USD: 50,
          BTC: 0.9000
        }
      }
    };

    const reducer = walletReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });
});
