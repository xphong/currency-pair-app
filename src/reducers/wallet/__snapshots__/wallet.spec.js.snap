// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet Reducer should handle RECEIVE_TRADE 1`] = `
Object {
  "balance": Object {
    "BTC": 0.9,
    "USD": 50,
  },
  "error": false,
  "isLoading": false,
  "success": true,
}
`;

exports[`Wallet Reducer should handle REQUEST_TRADE 1`] = `
Object {
  "balance": Object {
    "BTC": 0,
    "USD": 156.12,
  },
  "error": false,
  "isLoading": true,
  "success": false,
}
`;

exports[`Wallet Reducer should handle TRADE_ERROR 1`] = `
Object {
  "balance": Object {
    "BTC": 0,
    "USD": 156.12,
  },
  "error": "Insuccifient funds",
  "isLoading": false,
  "success": false,
}
`;

exports[`Wallet Reducer should return initial state 1`] = `
Object {
  "balance": Object {
    "BTC": 0,
    "USD": 156.12,
  },
  "error": false,
  "isLoading": false,
  "success": false,
}
`;
